Signal/src/ViewControllers/MediaPageViewController.swift

```diff
@@ -317,25 +317,51 @@ class MediaPageViewController: UIPageViewController, UIPageViewControllerDataSou
         let deleteAction = UIAlertAction(title: NSLocalizedString("TXT_DELETE_TITLE", comment: ""),
                                          style: .destructive) { _ in
                                             let deletedItem = currentViewController.galleryItem
-
-                                            if !self.sliderEnabled {
-                                                // In message details, which doesn't use the slider, so don't swap pages.
-                                            } else if let nextItem = mediaGalleryDataSource.galleryItem(after: deletedItem) {
-                                                self.setCurrentItem(nextItem, direction: .forward, animated: true)
-                                            } else if let previousItem = mediaGalleryDataSource.galleryItem(before: deletedItem) {
-                                                self.setCurrentItem(previousItem, direction: .reverse, animated: true)
-                                            } else {
-                                                // else we deleted the last piece of media, return to the conversation view
-                                                self.dismissSelf(animated: true)
-                                            }
-                                            mediaGalleryDataSource.delete(items: [deletedItem])
+                                            mediaGalleryDataSource.delete(items: [deletedItem], initiatedBy: self)
         }
         actionSheet.addAction(OWSAlerts.cancelAction)
         actionSheet.addAction(deleteAction)
 
         self.present(actionSheet, animated: true)
     }
 
+    // MARK: MediaGalleryDataSourceDelegate
+
+    func mediaGalleryDataSource(_ mediaGalleryDataSource: MediaGalleryDataSource, willDelete items: [MediaGalleryItem], initiatedBy: MediaGalleryDataSourceDelegate) {
+        Logger.debug("\(self.logTag) in \(#function)")
+
+        guard let currentItem = self.currentItem else {
+              owsFail("\(logTag) in \(#function) currentItem was unexpectedly nil")
+            return
+        }
+
+        guard items.contains(currentItem) else {
+            Logger.debug("\(self.logTag) in \(#function) irrelevant item")
+            return
+        }
+
+        // If we setCurrentItem with (animated: true) while this VC is in the background, then
+        // the next/previous cache isn't expired, and we're able to swipe back to the just-deleted vc.
+        // So to get the correct behavior, we should only animate these transitions when this
+        // vc is in the foreground
```

Comment:
This confounded me for a good while, and is the reason for the somewhat awkward `initiatedBy` param.